Update Pattern Library pages to use new color scheme

andy-armstrong opened the pull request:
### Description
 
[TNL-5135](https://openedx.atlassian.net/browse/TNL-5135)

This change updates edx-platform to use the latest release of the Pattern Library. In particular, this requires a number of color changes, as @clrux has greatly simplified the colors available (see https://github.com/edx/ux-pattern-library/pull/345). With Chris's help, I've picked what I believe to be the closest color in the new scheme. 

Please try out the sandbox and verify that I haven't really messed things up on your favorite pages!

### Sandbox
- [ ] https://andy-armstrong.sandbox.edx.org/dashboard

### Testing
- [ ] manual smoke test of various pages in LMS and Studio

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @bjacobel 
- [ ] Code review: @AlasdairSwan 
- [x] Accessibility review: @clrux
- [ ] Product review: @marcotuts 

FYI: @schenedx @cahrens 

### Post-review
- [ ] Squash commits

andy-armstrong commented:
@AlasdairSwan @schenedx There is an a11y failure for Programs that seems to be putting white on a light gray background. It would be great to get your help tomorrow in understanding where to look.

clrux commented on lms/static/certificates/sass/_components.scss:
```diff
@@ -96,27 +96,27 @@
 .wrapper-banner-user {
-    box-shadow: inset 0 rem(2) rem(2) 0 palette(grayscale, black-t);
-    background: palette(grayscale-cool, x-dark);
+    box-shadow: inset 0 rem(2) rem(2) 0 palette(grayscale, black);
```
I'm not sure we have a `black` key for grayscale. How about `dark` or `$black`?

andy-armstrong replied:
I think the original was pure black, so I've made it the same. I'm starting to think that we need variables for certificates colors so that it is easier to see the intent, and also to support theming. I'm not going to do that here though...

clrux commented on lms/static/certificates/sass/_components.scss:
```diff
@@ -161,7 +161,7 @@
 // main accomplishment
 .accomplishment-main {
-    background: palette(grayscale-cool, x-trans);
+    background: transparentize(palette(grayscale, back), 0.75);
```
Do we need to make this semi-transparent?

andy-armstrong replied:
The previous version used ``x-trans`` which makes it 75% transparent, so I was mimicking that.

clrux commented on lms/static/certificates/sass/_components.scss:
```diff
@@ -537,7 +537,7 @@
 // ------------------------------
 .footer-app {
-    border-top: rem(4) solid palette(grayscale, xx-trans);
+    border-top: rem(4) solid transparentize(palette(grayscale, base), 0.95);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

andy-armstrong replied:
I'm not lightening the gray, I'm making it very transparent. I don't know why that makes sense for a border, though. I need to work out how to view this on a sandbox so I can see the effect.

BTW, I find using the name ``back`` for anything other than a background color to be confusing. Why didn't we go with ``light``? It isn't clear reading the code what ``back`` or ``x-back`` mean.

AlasdairSwan replied:
Depending on the background below an element changing the transparency can alter the element's shade. I can see how this could be a useful way of defining a range of colours based on a single input colour, but I think we should only be using transparency when we want to be able to see through an element. Otherwise we cannot be sure what shade the colour will render as (2 of the same elements would be different shades if added to 2 different containers with different backgrounds).

bjacobel replied:
+1 for andy's comment about `back` (as opposed to `light`) -- I signed off on the code review that introduced `back` but seeing it used in context here I agree that `back` makes more sense. Something to think about as we expand this feature (as you discussed [here](https://github.com/edx/edx-platform/pull/13144#issuecomment-237218487), @clrux).

clrux commented on lms/static/certificates/sass/_components.scss:
```diff
@@ -556,7 +556,7 @@
     left: 0;
     z-index: z-index(front);
-    background-color: palette(grayscale, trans); /* dim the background */
+    background-color: transparentize(palette(grayscale, base), 0.5); /* dim the background */
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

andy-armstrong replied:
Again, my understanding is that this is making the gray transparent, not lightening it.

clrux commented on lms/static/certificates/sass/_config.scss:
```diff
@@ -7,12 +7,12 @@
-$cert-decorative-color-de-emphasized: palette(grayscale-cool, x-trans);
-$cert-decorative-color-emphasized: palette(grayscale-cool, mid-light);
+$cert-decorative-color: palette(grayscale, back);
+$cert-decorative-color-de-emphasized: transparentize(palette(grayscale, base), 0.75);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

clrux replied:
If a new shade of gray is absolutely necessary, I'm OK with adding one additional gray to our palette. But the goal is to keep it slim and simple.

andy-armstrong replied:
This is not a new shade of gray. This is a transparent version of the standard gray.

clrux commented on lms/static/sass/elements/_banners.scss:
```diff
@@ -10,7 +10,7 @@ $full-width-banner-margin: 20px;
         overflow: hidden;
         position: relative;
-        background: palette(grayscale, black-t);
+        background: palette(grayscale, black);
```
Is `black` a valid key? How about just `$black` here, or `dark`?

andy-armstrong replied:
Good catch. I've switched these two to ``$black``.

clrux commented on lms/static/sass/elements/_banners.scss:
```diff
@@ -19,7 +19,7 @@ $full-width-banner-margin: 20px;
             bottom: 0;
             left: 0;
-            background: palette(grayscale, black-t);
+            background: palette(grayscale, black);
```
Is `black` a valid key? How about just `$black` here, or `dark`?

clrux commented on lms/static/sass/elements/_course-card.scss:
```diff
@@ -138,7 +138,7 @@
   .upgrade-message,
   .certificate-status {
-    border-top: 1px solid palette(grayscale, x-trans);
+    border-top: 1px solid transparentize(palette(grayscale, base), 0.75);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

clrux commented on lms/static/sass/shared-v2/_footer.scss:
```diff
@@ -5,7 +5,7 @@
     @extend %ui-print-excluded;
     box-shadow: 0 -1px 5px 0 $shadow-l1;
-    border-top: 1px solid tint(palette(grayscale, light), 50%);
+    border-top: 1px solid tint(palette(grayscale, back), 50%);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

andy-armstrong replied:
Good point. I've switched this to use ``palette(grayscale, x-back)``.

clrux commented on lms/static/sass/shared-v2/_help-tab.scss:
```diff
@@ -12,16 +12,16 @@
-        background: transparentize(palette(grayscale, white-t), 0.25);
-        color: transparentize(palette(grayscale-cool, x-dark), 0.25);
+        background: transparentize($white, 0.25);
+        color: transparentize(palette(grayscale, dark), 0.25);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

clrux commented on lms/static/sass/shared-v2/_modal.scss:
```diff
@@ -122,7 +122,7 @@
 
         .modal-form-error {
-            background: palette(error, xx-trans);
+            background: transparentize(palette(error, back), 0.95);
```
Can we use one of the five values of gray? I'm hesitant to do this kind of thing because then we'll end up with our same fifty shades of grey. It looks like you're trying to lighten our base grey a smidge? What about using the `accent`, or `back`?

clrux commented:
@andy-armstrong Nice work here, thank you so very much for tackling this. Most of my comments were regarding the use of `black` as a grayscale key, which I don't think exists. And then the two or three values of gray that were being made transparent, I'm assuming to lighten them up a bit. Ideally, we'd make use of the various values prescribed in the pattern library, but if we really need another value, I'm OK adding one more shade. I just don't want us to get into another situation where we've got fifty different shades of grey, so keeping it slim is important.

andy-armstrong commented:
Thanks for the review @clrux. I'll clean up the usages of black as you suggest. 

For the transparent colors, I was nervous about just using a built-in fully-opaque color, since I assumed that the transparency was a desired effect. It looks like you removed all the transparent options from the Pattern Library, so in those cases, I used the ``transparentize`` function to generate a partially transparent color.

I think I need to find usages of each one of these and see how it looks in practice before and after my changes.

schenedx commented:
@andy-armstrong The style sheet you should look into is https://github.com/edx/edx-platform/blob/master/lms/static/sass/views/_program-details.scss#L1
If you need help to visually inspect this page, let @rlucioni, @mikedikan or me know because we just worked on the page.

AlasdairSwan commented:
@andy-armstrong it would be great if you could check each instance of ```transparentize``` being used so that it is only being used where the element should be semi-transparent, and not just to change the shade of the element's colour.

bjacobel commented:
LGTM :thumbsup:

bjacobel commented:
> verify that I haven't really messed things up on your favorite pages!

We have screenshot tests, right? That seems like the only way you could catch issues with this using automated tests - I'm sure they don't run on every page, though.

andy-armstrong commented:
@bjacobel There are only a couple of screenshot tests, so unfortunately I don't think they will help in this case. I don't know if there's a plan to have a comprehensive set. I could imagine they would be more trouble than they are worth, as every change we make would break them.

andy-armstrong commented:
@clrux @AlasdairSwan @bjacobel Thanks for the review, folks. I've removed all the ``transparentize`` calls except for one that explicitly wants to provide transparency, and I've fixed all the ``palette`` calls that had unsupported parameters. You can see my follow-on changes in my second commit if you just want to see what has changed.

I also have Programs working now in Studio on my sandbox where I found a few more bugs, which I've also addressed in the second commit. I'm updating the sandbox now to get all these changes so it should be ready around 4:15. For example, you can see my new program here:

https://studio-andy-armstrong.sandbox.edx.org/program/1

Please re-review when you have a chance. Thanks.

clrux commented:
👍 I think this is a good first step. I also think we'll need to continue evolving things as we solidify the UXPL palettes. For example, we may need to add one more level of gray (though I don't want to go beyond that), which means later we'll need to update this again.

I'm in talks with Christina and Alisan about making a `common` file (used for both LMS and Studio) that houses all of the UXPL variables. It'll allow us to use a single file to color updates, eventually eliminating the mess of colors in the variables file for each area.

andy-armstrong commented on package.json:
```diff
@@ -5,7 +5,7 @@
     "backbone-validation": "~0.11.5",
     "coffee-script": "1.6.1",
-    "edx-pattern-library": "0.16.1",
+    "edx-pattern-library": "0.16.3",
```
Reviewers: @bjacobel and I spoke about which version to include, and we decided upon 0.16.3. He will bring in 0.16.4 with his work to add a Pattern Library search box to the discussion board.

andy-armstrong commented:
@AlasdairSwan please re-review when you get a chance. I'd like to merge the PR this morning so that I can unblock @bjacobel's work.

AlasdairSwan commented:
👍
